packages/app/src/cli/models/extensions/schemas.ts

```diff
@@ -27,10 +27,19 @@ export const NewExtensionPointSchema = schema.define.object({
 export const OldExtensionPointsSchema = schema.define.array(schema.define.string()).default([])
 export const NewExtensionPointsSchema = schema.define.array(NewExtensionPointSchema)
 export const ExtensionPointSchema = schema.define.union([OldExtensionPointsSchema, NewExtensionPointsSchema])
+export const ApiVersionSchema = schema.define.union([
+  schema.define.literal('2022-10'),
+  schema.define.literal('2023-01'),
+  schema.define.literal('unstable'),
+  schema.define.literal('internal'),
+])
+
+export type ApiVersionSchemaType = schema.define.infer<typeof ApiVersionSchema>
 
 export const BaseUIExtensionSchema = schema.define.object({
   name: schema.define.string(),
   type: schema.define.string().default('ui_extension'),
+  apiVersion: ApiVersionSchema.optional(),
```

Comment:
Adding this here adds is as an optional field to all UI extensions.   I assume not all UI extensions would support this, so is this what we want? Instead, should we be adding it to just the `ui_extension` here? https://github.com/Shopify/cli/blob/add-app-home/packages/app/src/cli/models/extensions/ui-specifications/ui_extension.ts/#L11-L14

Long term it might not matter since all extensions will be moving to the `ui_extension`

Reply:
I did add the field to that schema, too, since it is used to parse the config during publishing. However, during development, in order to add that field on the dev server payload (https://github.com/Shopify/cli/blob/df63e49dd9a9c8763277b1467b6fdac65da65df8/packages/app/src/cli/services/dev/extension/payload.ts#L64), I had to add it here, too — in that context, the extension's specification aren't currently available to narrow the type better.

I found this whole thing to be pretty confusing to follow tbh, but I just went with it since a number of other fields on `BaseUIExtensionSchema` are also only relevant for a subset of extension types (`extensionPoints`, `capabilities`, `metafields`, `categories`).